api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,546 @@ package v1alpha1
+	//
+	// +optional
+	CRD *ExternalDNSCRDSourceOptions `json:"crd,omitempty"`
+
+	// OpenShiftRoute source currently has no unique configuration options.
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=Ignore;Allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the ExternalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// ServiceType determines what types of Service resources
+	// are watched by ExternalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP"
+	//
+	// One or more Service types can be specified, if desired.
+	//
+	// Note that using the "ClusterIP" service type will enable
+	// the ExternalDNS "--publish-internal-services" flag,
+	// which allows ExternalDNS to publish DNS records
+	// for ClusterIP services.
+	//
+	// If no service types are provided, ExternalDNS will be
+	// configured to create DNS records for LoadBalancer services
+	// only by default.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={"LoadBalancer"}
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	ServiceType []corev1.ServiceType `json:"serviceType,omitempty"`
+}
+
+type ExternalDNSCRDSourceOptions struct {
+	// Kind is the kind of the CRD
+	// source resource type to be
+	// consumed by ExternalDNS.
+	//
+	// e.g. "DNSEndpoint"
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Kind string `json:"kind"`
+
+	// Version is the API version
+	// of the given resource kind for
+	// ExternalDNS to use.
+	//
+	// e.g. "externaldns.k8s.io/v1alpha1"
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Version string `json:"version"`
+
+	// LabelFilter specifies a label filter
+	// to be used to filter CRD resource instances.
+	// Only one label filter can be specified on
+	// an ExternalDNS instance.
+	//
+	// +optional
+	LabelFilter map[string]string `json:"labelFilter,omitempty"`
+}
+
+// ExternalDNSStatus defines the observed state of ExternalDNS.
+type ExternalDNSStatus struct {
+	// Conditions is a list of operator-specific conditions
+	// and their status.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the most recent generation observed.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Zones is the configured zones in use by ExternalDNS.
+	Zones []string `json:"zones,omitempty"`
```

Comment:
Why do we track this if your policy only supports orphaning?

Reply:
to make it more obvious to administrators which DNS Zones ExternalDNS has been successfully able to bind to?
I suppose we could use generic conditions to express that...
@Miciah thoughts?